components/hab/src/command/pkg/uninstall.rs

```diff
@@ -0,0 +1,131 @@
+///     4a. If there are, we throw an error
+///     4b. If not, we delete the package
+/// 5. For each dependency we check if there are any packages which depend on it
+///     5a. If there are, we skip it
+///     5b. If there are not, we delete it from disk and the graph
+///
+pub fn start(ui: &mut UI, ident: &PackageIdent, fs_root_path: &Path, dryrun: bool) -> Result<()> {
+    // 1.
+    let pkg_install = PackageInstall::load(ident, Some(fs_root_path))?;
+    let ident = pkg_install.ident();
+    ui.begin(format!("Uninstalling package {}", &ident))?;
+
+    // 2.
+    let package_path = hfs::pkg_root_path(Some(&fs_root_path));
+    let packages = all_packages(&package_path)?;
+    let package_installs = load(&fs_root_path, &packages)?;
+
+    let mut graph = PackageGraph::new();
+    graph.build(&package_installs)?;
+
+    // 3.
+    let deps = graph.ordered_deps(&ident);
+
+    // 4.
+    match graph.count_rdeps(&ident) {
+        None => (), // package not in graph - m
+        Some(0) => {
+            let install = find_install(&package_installs, &ident).unwrap();
+            maybe_delete(ui, &install, dryrun)?;
+            graph.remove(&ident);
+        }
+        Some(c) => {
+            return Err(Error::CannotRemovePackage(ident.clone(), c));
+        }
+    }
+
+    // 5.
+    let mut count = 0;
+    for p in &deps {
+        match graph.count_rdeps(&p) {
+            None => (), // package not in graph - maybe panic!
+            Some(0) => {
+                let install = find_install(&package_installs, &p).unwrap();
+                if let Ok(true) = maybe_delete(ui, &install, dryrun) {
+                    count = count + 1;
+                }
+                graph.remove(&p);
+            }
+            Some(c) => {
+                ui.warn(format!(
+                    "Skipping package {}. It is a dependency of {} packages",
+                    p.clone(),
+                    c
+                ))?;
+            }
+        }
+    }
+
+    if dryrun {
+
+    } else {
+        ui.end(format!(
+            "Deleted package {} and {} dependencies",
+            &ident, count
+        ))?;
+    }
+    Ok(())
+}
+
+// JC: TODO: O(n)
+fn find_install(packages: &Vec<PackageInstall>, ident: &PackageIdent) -> Option<PackageInstall> {
+    for p in packages {
+        if p.ident() == ident {
+            return Some(p.clone());
+        }
+    }
+    return None;
+}
+fn maybe_delete(ui: &mut UI, install: &PackageInstall, dryrun: bool) -> Result<bool> {
+    let ident = install.ident();
+    if dryrun {
+        ui.status(Status::DryRunDeleting, &ident)?;
+        return Ok(false);
+    }
+
+    ui.status(Status::Deleting, &ident)?;
+    let pkg_dir = install.installed_path();
+    fs::remove_dir_all(&pkg_dir)?;
+    Ok(true)
+}
+
+fn load(fs_root_path: &Path, idents: &Vec<PackageIdent>) -> Result<Vec<PackageInstall>> {
+    let mut result = Vec::<PackageInstall>::new();
```

Comment:
Might as well use [`Vec::with_capacity`](http://doc.rust-lang.org/1.29.1/alloc/vec/struct.Vec.html#method.with_capacity) since we know the final size we expect.

Also, the `::<PackageInstall>` can be omitted; the compiler can infer the type of `Vec` you need from context in this case.

Also, since the return value of `all_packages` doesn't seem to be used other than as an argument to this function, why not call it internally and change this function to `load_all_packages` or something?

Reply:
Good point.  I had been going back and forward on if I should change all_packages into a `HashMap` to make the lookup above faster.  Probably simplest to leave as `Vec` for now and see if it becomes a bottleneck